capa/ghidra/capa_explorer.py

```diff
@@ -0,0 +1,242 @@
+import sys
+import json
+import logging
+import pathlib
+from typing import Dict, List
+from contextlib import suppress
+
+from ghidra.app.cmd.label import CreateNamespacesCmd
+from ghidra.program.model.symbol import Namespace, SourceType, SymbolType
+
+import capa
+import capa.main
+import capa.rules
+import capa.render.json
+import capa.ghidra.helpers
+import capa.features.extractors.ghidra.extractor
+
+logger = logging.getLogger("capa_explorer")
+
+
+class CAPADATA:
+    def __init__(self, namespace, scope, capability, locations, node, attack=None):
+        self.namespace = namespace
+        self.scope = scope
+        self.capability = capability
+        self.locations = locations
+        self.node = node
+        self.attack = attack
+
+    def recurse_node(self, node_dict):
+        """pull string data by recursing node dicts"""
+
+        if not node_dict:
+            # KeyError
+            # ex. {'type':'subscope', 'scope':'basic block'}
+            return ""
+
+        if isinstance(node_dict, int):
+            node_dict = hex(node_dict)
+
+        if isinstance(node_dict, str):
+            return node_dict
+
+        if "description" in node_dict:
+            return node_dict.get("description")
+        else:
+            return self.recurse_node(node_dict.get(node_dict.get("type")))
+
+    def create_namespace(self):
+        """create new ghidra namespace for each capa namespace"""
+
+        # handle rules w/o namespace -> capa lib rule
+        if self.namespace == "capa":
+            lib_str = Namespace.DELIMITER + "lib" + Namespace.DELIMITER
+            self.namespace = self.namespace + lib_str + self.capability.replace(" ", "-")
+        cmd = CreateNamespacesCmd(self.namespace, SourceType.USER_DEFINED)
+        cmd.applyTo(currentProgram())  # type: ignore [name-defined] # noqa: F821
+
+        return cmd.getNamespace()
+
+    def add_bookmark(self, addr, txt, category="CapaExplorer"):
+        """create bookmark at addr"""
+        currentProgram().getBookmarkManager().setBookmark(addr, "Info", category, txt)  # type: ignore [name-defined] # noqa: F821
+
+    def tag_functions(self):
+        """create function tags for capabilities"""
+
+        # self.locations[0] will always be the largest
+        # scoped offset yielded i.e. closest to an entrypoint
+        addr = toAddr(hex(self.locations[0]))  # type: ignore [name-defined] # noqa: F821
+        f = getFunctionContaining(addr)  # type: ignore [name-defined] # noqa: F821
+
+        # bookmark Mitre ATT&CK tactics @ function scope
+        if f:
+            f.addTag(self.capability)
+            if self.attack:
+                for item in self.attack:
+                    attack_txt = item.get("tactic") + Namespace.DELIMITER + item.get("id")
+                    self.add_bookmark(addr, attack_txt, "CapaExplorer::Mitre ATT&CK")
+
+    def bookmark_locations(self):
```

Comment:
Let's use more verbose local variable names in this function to improve readability e.g. `st` -> `symbol_table`